text/3074-cargo-path-bases.md

```diff
@@ -0,0 +1,294 @@
+```
+
+If a developer wishes to use such an auto-vendored dependency, a
+contract must be established with the build system about exactly where
+vendred dependencies will end up. And since that path may not be near
+the project's `Cargo.toml`, the user's `Cargo.toml` may end up with
+either an absolute path or a long relative path, both of which may not
+work on other hosts, and thus cannot be checked in (or must be
+overwritten in-place by the build system).
+
+The proposed mechanism aims to simplify both of these use-cases by
+introducing named "base" paths in the Cargo configuration
+(`.cargo/config.toml`). Path dependencies can then be given relative to
+those base path names, which can be set either by a local developer in
+their user-wide configuration (`~/.cargo/config.toml`), or by an
+external build system in a project-wide configuration file.
+
+This effectively makes a "group" of path dependencies available at some
+undisclosed location to `Cargo.toml`, which then only has to know the
+layout to path dependencies _within_ that directory, and not the path
+_to_ that directory.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+If you often use path dependencies that live in a particular location,
+or if you want to avoid putting long paths in your `Cargo.toml`, you can
+define path _base directories_ in your [Cargo
+configuration](https://doc.rust-lang.org/cargo/reference/config.html).
+Your path dependencies can then be specified relative to those
+directories.
+
+For example, say you have a number of projects checked out in 
+`/home/user/dev/rust/libraries/`. Rather than use that path in your
+`Cargo.toml` files, you can define it as a "base" path in
+`~/.cargo/config.toml`:
+
+```toml
+[base]
+dev = "/home/user/dev/rust/libraries/"
+```
+
+Now, you can specify a path dependency on a library `foo` in that
+directory in your `Cargo.toml` using
+
+```toml
+[dependencies]
+foo = { path = "foo", base = "dev" }
+```
+
+Base paths also work in the `[patch]` section in case there are
+dependencies you often want to override with a local checkout!
+
+Like with other path dependencies, keep in mind that both the base _and_
+the path must exist on any other host where you want to use the same
+`Cargo.toml` to build your project.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The Cargo configuration files receive a new configuration table, `base`.
+Each entry is a key-value pair where the key names a new path "base",
+and the value specifies the path to that path base. The path can be
+relative, in which case it is resolved relative to the containing
+configuration file.
+
+Dependency specifications gain a new field, `base`, which is expected to
+hold the name of an already-defined path base. The field only carries
+meaning for `path` dependencies (for now). To resolve a path dependency
```

Comment:
I know @joshtriplett had some thoughts around potentially using the same `base` keyword for both `path` and `git`, but if it's untenable then I wonder if we want them to diverge _more_. For example, we could have it be `within` or `relative-to` for `path` and keep `base` for URL-based dependencies (like `git`) since "base" is already a thing in HTML documents on the web.